### ISSUE TITLE
Map flickers on transition when viewState is stored in Redux

### ISSUE THREAD

**Issue by bergkvist** [visgl/deck.gl#4550]
#### Description
My problem: https://imgur.com/a/haBPnKB

When putting my viewState in Redux - and starting any kind of transition, I get a clear flicker, where the map will start moving, then return back to its starting position, before continuing the movement. I sometimes also get a warning like the one below when this happens:
```
[Violation] 'requestAnimationFrame' handler took 103ms
```

If this is some kind of a race condition, it is a very consistent one. The violation for requestAnimationFrame only happens sometimes.

Neither of these problems exist when I store the viewState in a "useState"-hook instead of Redux.

#### Repro Steps
CodeSandbox: https://codesandbox.io/s/inspiring-cohen-li1zy
 - Click the map, and hold the + key for some intense flickering.

#### Environment (please complete the following information):
 - Framework Version: [deck.gl 8.1.3]
 - Browser Version: [
    Chrome 80.0.3987.149 (Official Build) (64-bit),
    Chrome 80.0.3987.163 (Official Build) (64-bit)
  ]
 - OS: [Manjaro Linux, Windows 10]

#### Logs
Whenever the `onViewStateChange`-event fires, I log out the current viewState, and the next one like this: 
```jsx
<DeckGL
  layers={[]}
  viewState={viewState}
  controller={MapController}
  onViewStateChange={e => {
    console.table([viewState, e.viewState])
    setViewState(e.viewState)
  }}
>
  <StaticMap reuseMaps mapStyle={mapStyle} />
</DeckGL>
```

##### useState (everything works as expected)
![image](https://user-images.githubusercontent.com/410028/80632550-b9aafc80-8a57-11ea-8fc4-0072915b7df7.png)

##### Redux (notice the extra row of output)
![image](https://user-images.githubusercontent.com/410028/80634339-9e8dbc00-8a5a-11ea-87f9-773117583142.png)

**Comment by bergkvist** [comment]
I think the problem here could stem from how JavaScript (as a turn-based language) decides what to run next.

Essentially what we have here is a race condition that will only happen if "the turn" is given up in between:
 - emitting a new viewState (with a transition)
 - before receiving the viewState it just emitted

## Demonstration
Allowing the internal map handlers to run in between emitting a state and before receiving a new one can be accomplished with a setTimeout with 0ms.
```jsx
export default function WorldMap() {
  const [viewState, setViewState] = useState(initialViewState);
  return (
    <div>
      <DeckGL
        layers={[]}
        viewState={viewState}
        controller={MapController}
        onViewStateChange={e => {
          setTimeout(() => setViewState(e.viewState), 0);
        }}
      >
        <StaticMap
          reuseMaps
          mapboxApiAccessToken={mapboxApiAccessToken}
          mapStyle={mapStyle}
        />
      </DeckGL>
    </div>
  );
}
```

This completely messes up transitions. Way worse than the example using Redux.

## A potential solution
If the map never executes the onViewStateChange handler without first having received the latest viewState, then I believe that should fix the problem. 

This will prevent multiple viewState's from being out in the feedback loop at the same time. (which based on the log output in the Redux case seems to be causing the flicker)

**Comment by bergkvist** [comment]
Might be related to this issue: https://github.com/reduxjs/react-redux/issues/1298

**Comment by Pessimistress** [comment]
deck immediately dispatches an `onViewStateChange` when a new view state with transition is set. This is not an issue in non-React use cases, but React/Redux place a delay in between the update and the render.

One possible solution is to block the destination view state from rendering when the transition is first triggered. This will create a discrepancy between the React props and the camera for 1 frame, i.e. if you have other UI that updates to the longitude/latitude in the Redux store, they will go out of sync with the map during this frame.

Another solution is to add a `transitionTo` method on the `DeckGL` component. This will trigger `onViewStateChange` from the correct first frame.

**Comment by bergkvist** [comment]
Okay, so the `transitionTo`-method would mean that the first event here is never emitted?

![image](https://user-images.githubusercontent.com/410028/80743246-f55bca00-8b1c-11ea-98e5-ca6abe981b88.png)

**Comment by Pessimistress** [comment]
You are currently triggering the transition via an update to the Redux store:

-> redux store `viewState` updates with destination coordinates
-> `DeckGL` rerenders with new `viewState`
-> `onViewStateChange` is called
-> redux store `viewState` updates with intermediate coordinates
...

In the second proposal, you'll need to call the `transitionTo` API:
-> call `DeckGL`'s `ref.current.transitionTo(viewState)`
-> `onViewStateChange` is called
-> redux store `viewState` updates with intermediate coordinates

**Comment by bergkvist** [comment]
To avoid having to use refs, and to make the style a bit more declarative - what about doing something like:
```jsx
export default function WorldMap() {
  const [viewState, setViewState] = useState(initialViewState);
  const [transition, setTransition] = useState({})

  useEffect(() => {
    setTransition({
      inTransition: true,
      targetViewState: { longitude: 10, latitude: 62, zoom: 9 },
      durationLeft: 3000, // ms
    })
  }, [])

  return (
    <div>
      <DeckGL
        layers={[]}
        controller={MapController}

        // viewState doesn't know anything about transitions
        viewState={viewState}
        onViewStateChange={e => {
          setTimeout(() => setViewState(e.viewState), 0);
        }}

        // Instead, the transitions are handled/can be changed independently:
        transition={transition}
        onTransitionChange={setTransition}

      >
        <StaticMap
          reuseMaps
          mapboxApiAccessToken={mapboxApiAccessToken}
          mapStyle={mapStyle}
        />
      </DeckGL>
    </div>
  );
}
```

**Comment by bergkvist** [comment]
I guess an advantage of your first proposal is that it won't really change how the API is used from the outside? Meaning people wouldn't have to change their code. Is this correct? 

The trade-off here being the discrepancy between props and camera for one frame.

In the idea above, as well as your second proposal - people would need to modify their code.

**Comment by bergkvist** [comment]
I ended up creating this as a workaround: https://github.com/bergkvist/react-context-toolkit

React Context API doesn't seem to have [zero delays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop) between a dispatch and a rerender. (in contrast to react-redux)

**Comment by mayteio** [comment]
We use react context to manage this. Be wary of combining state and state updaters in one consumable hook to [avoid constant re-renders](https://kentcdodds.com/blog/how-to-use-react-context-effectively).

```js
// MapContext.js
import React from 'react';

const ViewportContext = createContext();
const SetViewportContext = createContext();

export const ViewportProvider = ({defaultViewport = {}, children}) => {
  const [viewport, setViewport] = useState(defaultViewport);

  return (
    <ViewportContext.Provider value={viewport}>
      <SetViewportContext.Provider value={setViewport}>
        {children}
      </SetViewportContext.Provider>
    </ViewportContext.Provider>
  )
}

/** Read the viewport from anywhere */
const useViewport = () => {
  const ctx = useContext(ViewportContext);
  if (!ctx) throw Error("Not wrapped in <ViewportProvider />.")
  return ctx;
}

/** Update the viewport from anywhere */
const useSetViewport = () => {
  const ctx = useContext(SetViewportContext);
  if (!ctx) throw Error("Not wrapped in <ViewportProvider />.")
  return ctx;
}
```
```
// attaching to deck
const viewport = useViewport();
const setViewport = useSetViewport();

<Deck 
  viewState={viewport} 
  onViewStateChange={({viewState}) => setViewport(viewState)}
/>
```

**Comment by Pessimistress** [comment]
This will be fixed in 8.5.0. You can try it right now with `8.5.0-alpha.1`.

**Comment by bergkvist** [comment]
@Pessimistress 
Thanks for looking into this

Tried it with 8.5.0-alpha.1 here: https://codesandbox.io/s/nice-spence-g5up8

You can try pressing plus (+) on your keyboard to trigger a zoom-out transition (previously this would zoom in).

I'm still seeing the same flicker as before.

**Comment by Pessimistress** [comment]
@bergkvist 

I need your help understanding what's going on in your app.

Let's add two log points in your code:

```jsx
export default function WorldMap() {
  const viewState = useSelector(selectViewState);
  const dispatch = useDispatch();
  console.log('rerender', viewState.zoom);
  return (
    <div>
      <DeckGL
        layers={[]}
        viewState={viewState}
        controller={MapController}
        onViewStateChange={e => {
          console.log('onViewStateChange', e.viewState.zoom);
          dispatch(setViewState(e.viewState));
        }}
      >
  ...
```

Now if you run the app, for every `onViewStateChange` call, the component is rerendered twice. 

If you remove `<React.StrictMode>`, then the double rendering is gone. However, you can see the following in the console:

```
onViewStateChange 7
rerender 7
onViewStateChange 6
rerender 7
onViewStateChange 6.053333333333333
rerender 6.053333333333333
```

Note that AFTER the second view state change is dispatched, the component is rendered again with the old view state. What is triggering this rerender?

**Comment by bergkvist** [comment]
@Pessimistress 
You've made some really interesting observations here. Taking your method/results into 3 dimensions of deck.gl, react-redux and StrictMode, we can see the following:

![image](https://user-images.githubusercontent.com/410028/108613095-ebc68a00-73ee-11eb-8522-f8bde4e4c17c.png)

![image](https://user-images.githubusercontent.com/410028/108613101-f84ae280-73ee-11eb-92a0-37c09cbe28d4.png)

Note that I've rounded the numbers to 2 decimal places, since I don't believe the decimals themselves have any significance here.

Some observations:
- It seems like using deck.gl v8.4.7 together with react-redux v7.2.1/v7.2.2 means the flicker doesn't happen regardless of React.StrictMode.
- Upgrading deck.gl from 8.4.7 to 8.5.0-alpha.1 reintroduces the visual flicker in strict mode for react-redux v7.2.1/v7.2.2.
- deck.gl v8.5.0-alpha.1 seems to make the "rerender with old value"-problem go away - although only in strict mode, but without actually fixing the visual flicker,

v7.2.1 of react-redux mentions some race-condition fixes, which are most likely relevant here: https://github.com/reduxjs/react-redux/releases/tag/v7.2.1

**Comment by Pessimistress** [comment]
That's actually great to know, thank you for digging into it. 8.5.0-alpha rewrote the DeckGL class component with function and hooks. I'll make sure it works in strict mode.

**Comment by Pessimistress** [comment]
Fix is included in `8.5.0-alpha.2`.

**Comment by bergkvist** [comment]
I just checked, and 8.5.0-alpha.2 works in StrictMode with react-redux 7.2.2 (no flickering). This is amazing!
